pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/LLCSegmentCompletionHandlers.java

```diff
@@ -276,8 +305,158 @@ public String segmentUpload(@QueryParam(SegmentCompletionProtocol.PARAM_INSTANCE
+    String tempMetadataDirStr = StringUtil.join("/", _controllerConf.getLocalTempDir(), segmentNameStr + METADATA_TEMP_DIR_SUFFIX);
+    File tempMetadataDir = new File(tempMetadataDirStr);
+
+    try {
+      Preconditions.checkState(tempMetadataDir.mkdirs(), "Failed to create directory: %s", tempMetadataDirStr);
+      // Extract metadata.properties from the metadataFiles.
+      if (!extractMetadataFromInputField(metadataFiles, tempMetadataDirStr, V1Constants.MetadataKeys.METADATA_FILE_NAME)) {
+        return null;
+      }
+      // Extract creation.meta from the metadataFiles.
+      if (!extractMetadataFromInputField(metadataFiles, tempMetadataDirStr, V1Constants.SEGMENT_CREATION_META)) {
+        return null;
+      }
+      // Load segment metadata
+      return new SegmentMetadataImpl(tempMetadataDir);
+    } catch (Exception e) {
+      LOGGER.error("Exception extracting and reading segment metadata for " + segmentNameStr, e);
+      return null;
+    } finally {
+      FileUtils.deleteQuietly(tempMetadataDir);
+    }
+  }
+
+  /**
+   *
+   * Extract a single file with name metaFileName from the input FormDataMultiPart and put it under the path
+   * tempMetadataDirStr + metaFileName.
+   * Return true iff the extraction and copy is successful.
+   */
+  private boolean extractMetadataFromInputField(FormDataMultiPart metadataFiles, String tempMetadataDirStr,
+                                                String metaFileName) {
+    FormDataBodyPart metadataFilesField = metadataFiles.getField(metaFileName);
+    Preconditions.checkNotNull(metadataFilesField, "The metadata input field %s does not exist.",
+            metaFileName);
+
+    try (InputStream metadataPropertiesInputStream = metadataFilesField.getValueAs(InputStream.class)) {
+      Preconditions.checkNotNull(metadataPropertiesInputStream, "Unable to parse %s from input.",
+              metaFileName);
+      java.nio.file.Path metadataPropertiesPath =
+              FileSystems.getDefault().getPath(tempMetadataDirStr, metaFileName);
+      Files.copy(metadataPropertiesInputStream, metadataPropertiesPath);
+      return true;
+    } catch (IOException e) {
+      LOGGER.error("Failed to copy metadata property file: {}", metaFileName, e);
+    }
+    return false;
+  }
+
+  /**
+   * Extract the segment metadata files from the tar-zipped segment file that is expected to be in the directory for the
+   * table.
+   * <p>Segment tar-zipped file path: segmentLocation.
+   * <p>We extract the metadata.properties and creation.meta into a temporary metadata directory:
+   * DATADIR/rawTableName/segmentName.metadata.tmp, and load metadata from there.
+   *
+   * @param segmentNameStr Name of the segment
+   * @param segmentLocation the location of the segment file.
+   * @return SegmentMetadataImpl if it is able to extract the metadata file from the tar-zipped segment file.
+   */
+  private SegmentMetadataImpl extractMetadataFromLocalSegmentFile(final String segmentNameStr,
```

Comment:
pls rename to extractMetadataFromSegmentFile. The segment file is in deep store, right?

Reply:
No. The segment file in this function is on local mounted FS. This should work on the current clients using local FS. I will modify this function so that it also works for clients on deep storage.

Reply:
Change done to make the method work for Deep storage. Note this function still download the segment file but it is not primary means to extract metadata now.